### PR TITLE
Redmine#7633: processexists() and findprocesses()

### PR DESCRIPTION
**Pull request by tzz**
See https://dev.cfengine.com/issues/7633

Example with body extension commented out:

```
bundle agent main
{
  methods:
      "test";

  vars:
      "test_state" data => bundlestate(test);
      "test_string" string => storejson(test_state);

  reports:
      "$(this.bundle): state of things = $(test_string)";
}

# body process_select myps
# {
#         pid => irange("1","1024");
#         process_result => "pid";
# }

bundle agent test
{
  vars:
      # "procs" data => findprocesses(".*", myps);
      "procs2" data => findprocesses(".*");
}
```

Output: all the processes on that system in a data container array holding key-value maps:

```
...
    {
      "line": "root     31292     2     0  0.0  0.0      0   0         0    1 Oct18  3-01:49:21 00:00:00 [kworker/3:0]",
      "pid": 31292
    },
...
```

### CODE REVIEW

**Comment by pasinskim** [comment]
trigger build

File: libpromises/evalfunction.c [hunk 1]
```diff
@@ -6888,6 +6889,47 @@ static int BuildLineArray(EvalContext *ctx, const Bundle *bundle,
 
 /*********************************************************************/
 
+static FnCallResult FnCallProcessExists(ARG_UNUSED EvalContext *ctx, ARG_UNUSED const Policy *policy, const FnCall *fp, const Rlist *finalargs)
+{
+    char *regex = RlistScalarValue(finalargs);
+
+    const bool context_mode = strcmp(fp->name, "processexists") == 0;
```
Comment (pasinskim) [inline libpromises/evalfunction.c]: Would be nice to have something more descriptive here like is_context_processexists or something similar.

File: libpromises/evalfunction.c [hunk 2]
```diff
@@ -6888,6 +6889,47 @@ static int BuildLineArray(EvalContext *ctx, const Bundle *bundle,
...
+        return FnFailure();
+    }
+
+    ProcessSelect ps;
+    Item *matched = SelectProcesses(procdata, regex, ps, false);
+    DeleteItemList(procdata);
+
+    if (context_mode)
+    {
+        return FnReturnContext(NULL != matched);
```
Comment (pasinskim) [inline libpromises/evalfunction.c]: We are having memory leak here. ```matched``` is allocated by ```SelectProcesses()``` but is never freed. Maybe something like ```bool ret = NULL != matched; DeleteItemList(matched); return ret;``` could be used instead.

File: libpromises/evalfunction.c [hunk 3]
```diff
@@ -6888,6 +6889,47 @@ static int BuildLineArray(EvalContext *ctx, const Bundle *bundle,
...
+    }
+
+    JsonElement *json = JsonArrayCreate(50);
```
Comment (pasinskim) [inline libpromises/evalfunction.c]: I know it's all over the places but I don't like "magic numbers" like this one. Maybe you can use some ```#define DEFAULT_PROCESS_EXISTS_JSON_CAPACITY 50``` or something similar?

File: libpromises/evalfunction.c [hunk 4]
```diff
@@ -6888,6 +6889,47 @@ static int BuildLineArray(EvalContext *ctx, const Bundle *bundle,
...
+    // we're in process gathering mode
+    for (Item *ip = matched; ip != NULL; ip = ip->next)
+    {
+        // we only have the ps line and PID
+        // TODO: this properly
```
Comment (pasinskim) [inline libpromises/evalfunction.c]: Can you extend the message a bit - like what "properly" means?

File: libpromises/evalfunction.c [hunk 5]
```diff
@@ -6888,6 +6889,47 @@ static int BuildLineArray(EvalContext *ctx, const Bundle *bundle,
...
+        JsonElement *pobj = JsonObjectCreate(4);
```
Comment (pasinskim) [inline libpromises/evalfunction.c]: Again, magic number. Why we are having ```JsonObjectCreate(4)```? We are adding just 2 elements there.

File: libpromises/evalfunction.c (hunk 1 above)
Comment (tzz) [inline libpromises/evalfunction.c]: I changed the variable name.  I prefer to check at the top so it's clear the C function has two modes.

File: libpromises/evalfunction.c (hunk 2 above)
Comment (tzz) [inline libpromises/evalfunction.c]: Got it, thanks.

File: libpromises/evalfunction.c (hunk 3 above)
Comment (tzz) [inline libpromises/evalfunction.c]: I am guessing how many elements this specific array will need, so it's really a guess of how many processes will be running.  I don't think it's a number that can be defined out.

File: libpromises/evalfunction.c (hunk 4 above)
Comment (tzz) [inline libpromises/evalfunction.c]: OK; done, thank you.

File: libpromises/evalfunction.c (hunk 5 above)
Comment (tzz) [inline libpromises/evalfunction.c]: Good point, changed.

**Comment by tzz** [comment]
OK, ready for another review.

**Comment by tzz** [comment]
If this is acceptable, I'll write acceptance tests and docs.

**Comment by kacfengine** [comment]
@pasinskim: ping

**Comment by pasinskim** [comment]
@tzz looks good now. Once there will be acceptance tests and corresponding documentation PR this is good to go. 
One small change that would be appreciated is 'Changelog: ' entry in the commit message.

**Comment by tzz** [comment]
@pasinskim PR rebased, changelog line added, and acceptance tests added.

**Comment by pasinskim** [comment]
trigger build

**Comment by pasinskim** [comment]
All checks passed.

**Comment by tzz** [comment]
@pasinskim I'll submit the docs in the next few days.  Thanks for reviewing and all the help!

**Comment by pasinskim** [comment]
@tzz this is nice and useful feature. Waiting for the docs.

**Pull request merged**